File: qif4j/src/main/java/com/github/squirrelgrip/qif4j/QifReaderListDecorator.java
package com.github.squirrelgrip.qif4j;

import java.io.PrintWriter;
import java.util.ArrayList;
import java.util.List;

public class QifReaderListDecorator {
	private final List<QifTransaction> transactions = new ArrayList<QifTransaction>();
	private final QifReader qifReader;

	public QifReaderListDecorator(QifReader qifReader) {
		this.qifReader = qifReader;
		qifReader.addTranasctionListener(new TransactionListener() {
			
			public void onTransaction(QifTransaction transaction) {
				transactions.add(transaction);
			}
		});
	}
	
	public List<QifTransaction> getTransactions() {
		qifReader.load();
		return transactions;
	}

	public void writeCsv(PrintWriter writer) {
		writeCsv(writer, QifReader.DEFAULT_DATE_FORMAT);
	}

	public void writeCsv(PrintWriter writer, String dateFormat) {
		writer.println("DATE,CATEGORY,PAYEE,TOTAL,MEMO");
		for (QifTransaction transaction : getTransactions()) {
			if (transaction instanceof QifCashTransaction) {
				writer.println(((QifCashTransaction) transaction).toCsv(dateFormat));
			}
		}
	}


}

File: qif4k/src/main/kotlin/com/github/squirrelgrip/qif/QifReaderListDecorator.kt
package com.github.squirrelgrip.qif

import java.io.PrintWriter
import java.util.*

class QifReaderListDecorator(private val qifReader: QifReader) {
    private val _transactions: MutableList<QifTransaction> = mutableListOf()

    init {
        qifReader.addTranasctionListener(object : TransactionListener {
            override fun onTransaction(transaction: QifTransaction) {
                _transactions.add(transaction)
            }
        })
        qifReader.load()
    }

    val transactions: List<QifTransaction>
        get() = _transactions.toList()

    @JvmOverloads
    fun writeCsv(writer: PrintWriter, dateFormat: String? = QifReader.DEFAULT_DATE_FORMAT) {
        writer.println("DATE,CATEGORY,PAYEE,TOTAL,MEMO")
        transactions.forEach {
            if (it is QifCashTransaction) {
                writer.println(it.toCsv(dateFormat))
            }
        }
    }

}